### PR TITLE
SAI Generic Programmable Extensions header

### PR DESCRIPTION
**Pull request by ravi861**
Signed-off-by: Ravi Vantipalli <ravi.vantipalli@intel.com>

This PR adds the header files and relevant meta code changes to SAI. The proposal is added and reviewed in PR #1551.

### CODE REVIEW

File: inc/saitypes.h [hunk 1]
```diff
@@ -989,6 +990,35 @@ typedef struct _sai_segment_list_t
...
+ * sai_attr_value_type - Data type of the attribute
+ * brief - Optional description of the field
+ * sai_attr_flags - Optional Usage flags for the field
+ * allowed_object_types - If data type is OID, then this is the list of object types allowed as data
+ */
+typedef struct _sai_json_t
+{
+    /** String in JSON format */
+    sai_s8_list_t json;
+} sai_json_t;
```
Comment (kcudnik) [inline inc/saitypes.h:1020]: why we need special type if it's just wrapper ? and not use sai_s8_list_t directly?

**Review by kcudnik — CHANGES_REQUESTED** [review]
check comments

File: inc/saitypes.h (hunk 1 above)
Comment (ravi861) [reply inc/saitypes.h:1020]: Having a new SAI_ATTR_VALUE_TYPE for json is very useful in generic code like sonic-sairedis. The JSON is allowed to carry other SAI datatypes like OID, IP address, etc and knowing that the attribute is a JSON helps with parsing it.
If I use straight, sai_s8_list_t, I don't think there is a way to distinguish between a simple string and a JSON string.
Would it help if we use a count/list within sai_json_t directly?

Comment (kcudnik) [reply inc/saitypes.h:1020]: no, lets leave this as is then

**Review by kcudnik — APPROVED** [review]

**Comment by reshmaintel** [comment]
@JaiOCP @marian-pritsak @rck-innovium Please do complete this review. Thanks.

**Comment by rlhui** [comment]
@JaiOCP @marian-pritsak @rck-innovium - please review/approve?  Thanks.

**Comment by rlhui** [comment]
@ravi861  - please add description of this PR and can we not use "GPE" in title but expand it?

File: inc/saigenericprogrammable.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    SAI_GENERIC_PROGRAMMABLE_ATTR_OBJECT_NAME = SAI_GENERIC_PROGRAMMABLE_ATTR_START,
+
+    /**
+     * @brief JSON string carrying HW block entry information
+     *
+     * @type sai_json_t
+     * @flags CREATE_AND_SET
+     * @default vendor
+     */
+    SAI_GENERIC_PROGRAMMABLE_ATTR_ENTRY,
```
Comment (rck-innovium) [inline inc/saigenericprogrammable.h:61]: If not already done, please add API examples that cover:

1) creating multiple entries of the same object name. For example, creating multiple VIP entries.


2) Updating a single attribute of an entry. For example, changing the nexthop-id of a VIP entry.

**Review by rck-innovium — APPROVED** [review]

**Review by marian-pritsak — APPROVED** [review]

**Comment by rlhui** [comment]
Can we please add more description to the PR. Currently it only has "sign-off by ..."

File: inc/saigenericprogrammable.h (hunk 2 above)
Comment (rlhui) [reply inc/saigenericprogrammable.h:61]: Hi, I don't see example of "creating multiple VIP entries" in https://github.com/opencomputeproject/SAI/blob/19ba5116264829d3a7b933212ab6bd0d977a3a55/doc/SAI_Proposal_Generic_Extensions.md, and also changing the entry?

Comment (ravi861) [reply inc/saigenericprogrammable.h:61]: Adding multiple different entries is the same as the single create example. The same invocation as existing example would work with modified IP and nexthop information.
I added an code example for update scenario also.

**Review by rlhui — APPROVED** [review]

**Pull request merged**